### PR TITLE
Add a job to deploy contracts on testnet

### PR DESCRIPTION
**Pull request by michalinacienciala**
This commit/PR adds a deployment of coverage-pools contracts on a
testnet (name of the testnet on which deployment will take place will
depend on the value of `environment` input passed when invoking
`workflow_dispatch` event). Before deploy script will run, we'll update
dependencies to upstream packages (`keep-core` and `tbtc` contracts are
needed for the deployment of `coverage-pools` contracts.)
Once contracts get deployed, we will push them to Tenderly (only
`ropsten` env curently supported) and we will publish npm package (which
will be later pulled by the Token Dashboard - it's not the scope of this
PR).
Last step in the job is an execution of `notify-workflow-completed`
action which notifies CI about successful run of the workflow and as a
result triggers deploy of any downstream modules specified in the
`ci/onfig.config.json` file.
As part of this PR we also change the conditions of execution of
`contracts-lint` and `contracts-slither` jobs - those jobs will no
longer be executed for `workflow_dispatch` event.

Depends on:
https://github.com/keep-network/coverage-pools/pull/165
https://github.com/keep-network/coverage-pools/pull/168

TODO:

- [x] Remove `--dry-run` in publish step before merging
- [x] Configure an access to `TENDERLY_TOKEN` from `coverage-pools` repo

### CODE REVIEW

File: .github/workflows/contracts.yml [hunk 1]
```diff
@@ -48,8 +60,85 @@ jobs:
...
+      - uses: actions/checkout@v2
+
+      - name: Load environment variables
+        uses: keep-network/load-env-variables@v1
+        with:
+          environment: ${{ github.event.inputs.environment }}
+
+      - uses: actions/setup-node@v2
+        with:
+          node-version: "12.x"
```
[the remark this answers is not in the document]
Comment (michalinacienciala) [reply .github/workflows/contracts.yml]: Fixed in the `coverage-pools` workflow. For workflows in other repositories, I will switch to `14.x` when working on some other changes.

**Comment by michalinacienciala** [comment]
We've got both PRs (#165, #168) required for the workflow success merged to main. I've merged latest code from main to the #155's feature branch (`add-cov-pools-to-ci`) and run a [deployment job](https://github.com/keep-network/coverage-pools/runs/3171212441?check_suite_focus=true).

It's failing on `hardhat deploy --export export.json --network ropsten` step with following error:

```
failed to get chainId, falling back on net_version...
```
I see that `chainId` is configured for `ropsten` in `hardhat.config.ts`, so I'm not sure why we're getting this error. @nkuba, any ideas?

FYI, I run the workflow providing upstream builds that were executed yesterday.

File: .github/workflows/contracts.yml [hunk 2]
```diff
@@ -48,14 +60,89 @@ jobs:
...
+            tbtc-solidity-version = github.com/keep-network/tbtc/solidity#version
+
+      - name: Resolve latest contracts
+        run: |
+          yarn add \
+            @keep-network/keep-core@${{ steps.upstream-builds-query.outputs.keep-core-solidity-version }} \
+            @keep-network/tbtc@${{ steps.upstream-builds-query.outputs.tbtc-solidity-version }}
+
+      - name: Deploy contracts
+        run: yarn deploy --network ${{ github.event.inputs.environment }}
```
Comment (nkuba) [inline .github/workflows/contracts.yml:99]: We need to make sure two env variables are set: `CHAIN_API_URL` and `CONTRACT_OWNER_ACCOUNT_PRIVATE_KEY`.

https://github.com/keep-network/coverage-pools/blob/2c290477c34d460dfb2186cfd7bf1f0d6b517672/hardhat.config.ts#L42-L44

Comment (michalinacienciala) [reply .github/workflows/contracts.yml:103]: I've added passing of env variables in the deploy step (see b4d31fafeeb7c9597e490b33a1ced6cc811a2ad1) , but something is still off.  I'm still getting the `failed to get chainId, falling back on net_version...` error (see https://github.com/keep-network/coverage-pools/runs/3172830878?check_suite_focus=true).
I'll try to find the root cause.

File: .github/workflows/contracts.yml [hunk 3]
```diff
@@ -48,14 +60,93 @@ jobs:
...
+        run: |
+          yarn add \
+            @keep-network/keep-core@${{ steps.upstream-builds-query.outputs.keep-core-contracts-version }} \
+            @keep-network/tbtc@${{ steps.upstream-builds-query.outputs.tbtc-contracts-version }}
+
+      - name: Deploy contracts
+        env:
+          CHAIN_API_URL: ${{ secrets.KEEP_TEST_ETH_HOSTNAME }}
+          CONTRACT_OWNER_ACCOUNT_PRIVATE_KEY: |
+            ${{ secrets.KEEP_TEST_ETH_CONTRACT_OWNER_PRIVATE_KEY }}
```
Comment (nkuba) [inline .github/workflows/contracts.yml:102]: Can we try if the below suggestion fixes the issue with depoyment?
```suggestion
          CONTRACT_OWNER_ACCOUNT_PRIVATE_KEY: ${{ secrets.KEEP_TEST_ETH_CONTRACT_OWNER_PRIVATE_KEY }}
```

**Comment by michalinacienciala** [comment]
@nkuba, I removed the `--dry-run` from the `npm publish`. The only remaining piece of work needed for the workflow passing is configuring the access to the `TENDERLY_TOKEN`. Once added, workflow can be verified by re-runing of https://github.com/keep-network/coverage-pools/runs/3180388501?check_suite_focus=true workflow.

**Comment by michalinacienciala** [comment]
> The only remaining piece of work needed for the workflow passing is configuring the access to the `TENDERLY_TOKEN`. Once added, workflow can be verified by re-runing of https://github.com/keep-network/coverage-pools/runs/3180388501?check_suite_focus=true workflow.

After configuring the secret, we can authenticate in Tenderly, but pushing contracts to Tenderly fails (see [the the workflow with the failure](https://github.com/keep-network/coverage-pools/actions/runs/1074989319)). This is not a blocker for this PR and will be investigated/fixed as part of https://keepnetwork.atlassian.net/browse/CP-52 task.

@nkuba, ready for review.

File: tenderly.yaml [hunk 4]
```diff
@@ -0,0 +1,7 @@
+account_id: d9a948cf-8667-43df-acb0-ce194f6dda7a
+projects:
+  thesis/keep:
+    networks:
+      - "1" # mainnet
+  thesis/keep-test:
+    - "3" # ropsten
```
Comment (nkuba) [inline tenderly.yaml:7]:
```suggestion
  thesis/keep-test:
    networks:
      - "3" # ropsten
```

**Review by nkuba — CHANGES_REQUESTED** [review]

File: .github/workflows/contracts.yml [hunk 5]
```diff
@@ -48,14 +60,92 @@ jobs:
       - name: Deploy contracts
         run: yarn deploy
 
+  contracts-deployment-testnet:
+    needs: [contracts-build-and-test]
+    if: github.event_name == 'workflow_dispatch'
```
[the remark this answers is not in the document]
Comment (michalinacienciala) [reply .github/workflows/contracts.yml:65]: Hmm, I think there may be cases when we'll want to run the deployment job on other branches than `main`. And if we'll want to run the `contracts-build-and-test` job on some branch without deploying, we can do that by creating draft PR on that feature branch or if PR already exists, committing to the branch or re-running already executed runs.

**Review by nkuba — APPROVED** [review]

**Pull request merged**